### PR TITLE
[SPARK-19690][SS] Join a streaming DataFrame with a batch DataFrame which has an aggregation may not work

### PR DESCRIPTION
**Pull request by uncleGen**
## What changes were proposed in this pull request?

`StatefulAggregationStrategy` should check logicplan is streaming or not

Test code:

```
case class Record(key: Int, value: String)
val df = spark.createDataFrame((1 to 100).map(i => Record(i, s"value_$i"))).groupBy("value").count
val lines = spark.readStream.format("socket").option("host", "localhost").option("port", "9999").load 
val words = lines.as[String].flatMap(_.split(" ")) 
val result = words.join(df, "value")
```

before pr:

```
== Physical Plan ==
*Project [value#13, count#19L]
+- *BroadcastHashJoin [value#13], [value#1], Inner, BuildRight
   :- *Filter isnotnull(value#13)
   :  +- *SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, input[0, java.lang.String, true], true) AS value#13]
   :     +- MapPartitions <function1>, obj#12: java.lang.String
   :        +- DeserializeToObject value#5.toString, obj#11: java.lang.String
   :           +- StreamingRelation textSocket, [value#5]
   +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, string, true]))
      +- *HashAggregate(keys=[value#1], functions=[count(1)])
         +- StateStoreSave [value#1], OperatorStateId(<unknown>,0,0), Append, 0
            +- *HashAggregate(keys=[value#1], functions=[merge_count(1)])
               +- StateStoreRestore [value#1], OperatorStateId(<unknown>,0,0)
                  +- *HashAggregate(keys=[value#1], functions=[merge_count(1)])
                     +- Exchange hashpartitioning(value#1, 200)
                        +- *HashAggregate(keys=[value#1], functions=[partial_count(1)])
                           +- *Project [value#1]
                              +- *Filter isnotnull(value#1)
                                 +- LocalTableScan [key#0, value#1]
```

after pr:

```
== Physical Plan ==
*Project [value#13, count#19L]
+- *BroadcastHashJoin [value#13], [value#1], Inner, BuildRight
   :- *Filter isnotnull(value#13)
   :  +- *SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, input[0, java.lang.String, true], true) AS value#13]
   :     +- MapPartitions <function1>, obj#12: java.lang.String
   :        +- DeserializeToObject value#5.toString, obj#11: java.lang.String
   :           +- StreamingRelation textSocket, [value#5]
   +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, string, true]))
      +- *HashAggregate(keys=[value#1], functions=[count(1)])
         +- Exchange hashpartitioning(value#1, 200)
            +- *HashAggregate(keys=[value#1], functions=[partial_count(1)])
               +- *Project [value#1]
                  +- *Filter isnotnull(value#1)
                     +- LocalTableScan [key#0, value#1]
```

## How was this patch tested?

add new unit test.

### CODE REVIEW

**Comment by uncleGen** [comment]
cc @zsxwing

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 1]
```diff
@@ -535,7 +535,8 @@ case class Range(
 case class Aggregate(
     groupingExpressions: Seq[Expression],
     aggregateExpressions: Seq[NamedExpression],
-    child: LogicalPlan)
+    child: LogicalPlan,
+    stateful: Boolean = false)
   extends UnaryNode {
```
Comment (uncleGen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: `stateful` indicates if the aggregate is base on streaming or batch, resolved by `ResolveStatefulAggregate` rule

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 2]
```diff
@@ -393,6 +393,17 @@ case class PreprocessTableInsertion(conf: SQLConf) extends Rule[LogicalPlan] {
...
 
 /**
+ * A rule to do check if the aggregate is stateful.
+ */
+class ResolveStatefulAggregate() extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case agg @ Aggregate(groupingExpressions, aggregateExpressions, child, _)
+      if agg.isStreaming =>
+        Aggregate(groupingExpressions, aggregateExpressions, child, stateful = true)
+  }
```
Comment (uncleGen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: resolve one aggregate, determine statefule or not.

**Comment by zsxwing** [comment]
Thanks for doing this. I'm wondering if you can fix `isStreaming` instead. We added it to be able to distinguish batch and streaming dataframes. However, it doesn't work for batch DFs in a streaming query. My initial thought is making sure all streaming Source should return a DataFrame whose `isStreaming` must return `true`. Could you try to resolve this issue in this direction?

**Comment by uncleGen** [comment]
@zsxwing got it

**Comment by uncleGen** [comment]
retest this please.

**Comment by uncleGen** [comment]
working on unit test failure

**Comment by uncleGen** [comment]
retest this please.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 3]
```diff
@@ -989,7 +989,8 @@ class FileStreamSourceSuite extends FileStreamSourceTest {
       withSQLConf(
         SQLConf.FILE_SOURCE_LOG_COMPACT_INTERVAL.key -> "2",
         // Force deleting the old logs
-        SQLConf.FILE_SOURCE_LOG_CLEANUP_DELAY.key -> "1"
+        SQLConf.FILE_SOURCE_LOG_CLEANUP_DELAY.key -> "1",
+        SQLConf.UNSUPPORTED_OPERATION_CHECK_ENABLED.key -> "false"
       ) {
```
Comment (uncleGen) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: Close the "UNSUPPORTED_OPERATION_CHECK_ENABLED", as `Source.getBatch` returns DF whose `isStreaming` is true.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 4]
```diff
@@ -1119,11 +1119,16 @@ case class DecimalAggregates(conf: CatalystConf) extends Rule[LogicalPlan] {
...
       val projection = new InterpretedProjection(projectList, output)
       projection.initialize(0)
-      LocalRelation(projectList.map(_.toAttribute), data.map(projection))
+      if (lr.isStreaming) {
+        LocalRelation(projectList.map(_.toAttribute), data.map(projection))
+          .setIncremental()
+      } else {
+        LocalRelation(projectList.map(_.toAttribute), data.map(projection))
+      }
   }
```
Comment (uncleGen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: In a streaming query, we will transfrom stream source to a batch `LocalRelation` whose `isStreaming` is true, so we should keep new LocalRelation's `isStreaming` is true in this rule.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 5]
```diff
@@ -231,8 +231,9 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
       case EventTimeWatermark(columnName, delay, child) =>
         EventTimeWatermarkExec(columnName, delay, planLater(child)) :: Nil
 
-      case PhysicalAggregation(
-        namedGroupingExpressions, aggregateExpressions, rewrittenResultExpressions, child) =>
+      case agg @ PhysicalAggregation(
+        namedGroupingExpressions, aggregateExpressions, rewrittenResultExpressions, child)
+        if agg.isStreaming =>
```
Comment (uncleGen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Apply this strategy only if the logicplan is streaming.

**Comment by uncleGen** [comment]
retest this please.

**Comment by uncleGen** [comment]
cc @zsxwing

**Comment by uncleGen** [comment]
\cc @zsxwing

**Comment by uncleGen** [comment]
ping @zsxwing

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 6]
```diff
@@ -31,6 +31,8 @@ abstract class LogicalPlan extends QueryPlan[LogicalPlan] with Logging {
 
   private var _analyzed: Boolean = false
 
+  private var _incremental: Boolean = false
```
Comment (zsxwing) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala]: Adding it here will break `sameResult`, `equals` and other methods. Could you add a new parameter to the constructor of `LogicalRelation` and `LogicalRDD` instead?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Source.scala [hunk 7]
```diff
@@ -57,10 +60,31 @@ trait Source  {
...
+        logicalRDD.dataFromStreaming = true
+        logicalRDD
+      case logicalRelation @ LogicalRelation(_, _, _, false) =>
+        logicalRelation.dataFromStreaming = true
+        logicalRelation
+      case localRelation @ LocalRelation(_, _, false) =>
+        localRelation.dataFromStreaming = true
+        localRelation
+    }
+
```
Comment (uncleGen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Source.scala]: add a new parameter `dataFromStreaming ` to the constructor of LogicalRelation, LogicalRDD and LocalRelation. `dataFromStreaming ` indicate if this relation comes from a streaming source. In a streaming query, stream relation will be cut into a series of batch relations.

**Comment by uncleGen** [comment]
retest this please.

**Comment by uncleGen** [comment]
retest this please.

**Comment by srowen** [comment]
This is marked as "Critical" for 2.1.1, but I'm not clear it's a regression or that urgent?

**Comment by HyukjinKwon** [comment]
@uncleGen is this still active?

**Comment by uncleGen** [comment]
@HyukjinKwon Sorry! Busy for this period of time. Let me resolve this conflict.

**Comment by HyukjinKwon** [comment]
Yea, I just wanted to check if it is in progress in any way. Thanks for your input.

**Pull request closed without merging**